src/components/figures/chart/chartJs/chartjs_funnel_chart.ts

```diff
@@ -0,0 +1,121 @@
+  TooltipPositionerFunction,
+} from "chart.js";
+
+export class FunnelChartController extends window.Chart?.BarController {
+  static id = "funnel";
+  static defaults = { ...window.Chart?.BarController.defaults, dataElementType: "funnel" };
+
+  updateElements(rects, start, count, mode) {
+    super.updateElements(rects, start, count, mode);
+    for (let i = start; i < start + count; i++) {
+      const rect = rects[i];
+      this.updateElement(rect, i, { nextElementWidth: rects[i + 1]?.width || 0 }, mode);
+    }
+  }
+}
+
+/**
+ * Similar to a bar chart element, but it's a trapezoid rather than a rectangle. The top is of width
+ * `width`, and the bottom is of width `nextElementWidth`.
+ */
+export class FunnelChartElement extends window.Chart?.BarElement {
+  static id = "funnel";
+
+  draw(ctx: CanvasRenderingContext2D) {
+    ctx.save();
+
+    const { x, y, width, height, nextElementWidth, base, options } = this.getProps([
+      "x",
+      "y",
+      "width",
+      "height",
+      "nextElementWidth",
+      "base",
+      "options",
+    ]) as any;
+    const offset = (width - nextElementWidth) / 2;
+
+    const startX = Math.min(x, base);
+    const startY = y - height / 2;
+
+    ctx.fillStyle = options.backgroundColor;
+
+    ctx.beginPath();
+    ctx.moveTo(startX, startY);
+    ctx.lineTo(startX + width, startY);
+    ctx.lineTo(startX + width - offset, startY + height);
+    ctx.lineTo(startX + offset, startY + height);
+    ctx.closePath();
+    ctx.fill();
+    if (options.borderWidth) {
+      ctx.strokeStyle = options.borderColor;
+      ctx.lineWidth = options.borderWidth;
+      ctx.stroke();
+    }
+
+    ctx.restore();
+  }
+
+  /** Check if the mouse is inside the trapezoid */
+  inRange(mouseX: number, mouseY: number, useFinalPosition: boolean) {
+    const { x, y, width, height, nextElementWidth, base } = this.getProps(
+      ["x", "y", "width", "height", "nextElementWidth", "base"],
+      useFinalPosition
+    ) as any;
+
+    const startX = Math.min(x, base);
+    const startY = y - height / 2;
+    if (mouseY < startY || mouseY > startY + height) {
+      return false;
+    }
+    const offset = (width - nextElementWidth) / 2;
+    const left = startX + (offset * (mouseY - startY)) / height;
+    const right = startX + width - (offset * (mouseY - startY)) / height;
+    if (mouseX < left || mouseX > right) {
+      return false;
+    }
+
+    return true;
+  }
+}
+
+const funnelTooltipPositioner: TooltipPositionerFunction<"funnel"> = function (elements) {
+  if (!elements.length) {
+    return { x: 0, y: 0 };
+  }
+
+  const element = elements[0].element as BarElement;
+  const { x, y, base, width, height } = element.getProps(["x", "y", "width", "height", "base"]);
+  const startX = Math.min(x, base);
+  const startY = y - height / 2;
+
+  return {
+    x: startX + (width * 2) / 3,
+    y: startY + height / 2,
+  };
+};
+
+window.Chart.Tooltip.positioners.funnelTooltipPositioner = funnelTooltipPositioner;
```

Comment:
how does chartjs knows it must use `funnelTooltipPositioner` ? is it a convention `<chart type>TooltipPositioner` ?

Reply:
![image](https://github.com/user-attachments/assets/2444890d-d89b-4b71-8f55-8b7792ca526c)

We just give it as argument of the tooltip 😅